pytorch_lightning/trainer/connectors/logger_connector/result.py

```diff
@@ -74,6 +74,11 @@ def _generate_sync_fn(self) -> None:
         # ignore typing, bad support for `partial`: mypy/issues/1484
         self._fn: Callable = partial(fn, reduce_op=self.op, group=self.group)  # type: ignore [arg-type]
 
+    def __deepcopy__(self, memo: dict) -> "_Sync":
```

Comment:
Can you add a comment about why this is necessary?

Reply:
In trying to find a satisfactory explanation, I discovered this change was just patching a symptom. The setter for `_Metadata.sync` alters the reduce operation for the `_Sync` object but does not regenerate its `_fn` attribute ([the problematic setter invocation](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/result.py#L429)). This (presumed) bug was masked in previous versions by [the call to `self.register_key`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/logger_connector/result.py#L433), which implicitly created a new `_Sync` object and regenerated `_fn` due to the way dataclasses were handled in `apply_to_collection`. Changing to a (deep)copy (forced by the inability to infer default values for `InitVar` constructor arguments) meant that this regeneration no longer occurred and that the `partial`ly applied `_fn` retained a `None` reduce op. 

I will be pushing a more targeted fix that eschews the custom `__deepcopy__` shortly.